common/event_emitter.go

```diff
@@ -164,12 +171,22 @@ func (e *BaseEventEmitter) sync(fn func()) {
 	<-done
 }
 
-func (e *BaseEventEmitter) emit(event string, data interface{}) {
+func (e *BaseEventEmitter) emit(logger *log.Logger, event string, data interface{}) {
 	emitEvent := func(eh eventHandler) {
-		select {
-		case eh.ch <- Event{event, data}:
-		case <-eh.ctx.Done():
-			// TODO: handle the error
+		for {
+			select {
+			case eh.ch <- Event{event, data}:
+				return
+			case <-eh.ctx.Done():
+				// TODO: handle the error
+				return
+			case <-time.After(time.Second):
+				// If this is printed then a handler is deadlocked on something
+				// and the channel (which should be buffered) is full. It might
+				// be that by increasing the buffer size on the channel will help
+				// alleviate the problem.
+				logger.Error("emit timed out waiting for handler to read message")
```

Comment:
Should we then cut this short here (`return` without `for`)? Now it retries, is it expected?

Reply:
Yes, i'll add a `return` if we're all happy with this change -- it's currently useful as a tool to debug an issue with pprof 🙂